### PR TITLE
Python spikes

### LINKED ISSUE #764: Python wrapper: spikes
Wrapper for
* `arb::spike` type
* helper for recording spike trains
* example of recording and printing/plotting spikes

### PR DESCRIPTION
**Pull request by bcumming**
Support for recording spikes generated by a simulation in the Python wrapper
* Implement a `spike_recorder` that holds a shared pointer to a `std::vector` of spikes, and a callback for the `arb::simulation` spike recording API.
* Add `python/example/ring.py` that creates a ring network, then records and prints spikes.
* Some fixes to get the full `recipe` -> `domain_decomposition` -> `simulation` -> `spikes` workflow to work
  * always use default `global_parameters`: user customization of global parameters for cable cells can wait until the ion species interface is finished.
  * change the Python recipe interface for `recipe::connections_on` to use `pybind11::objects` because of shim.
* Some small improvements to error and help messages.

Fixes #764

### CODE REVIEW

File: python/recipe.cpp [hunk 1]
```diff
@@ -112,6 +118,32 @@ struct cell_connection_shim {
...
+    auto pycons = impl_->connections_on(gid);
+    std::vector<arb::cell_connection> cons;
+    cons.reserve(pycons.size());
+    for (unsigned i=0; i<pycons.size(); ++i) {
+        const auto& c = pycons[i];
+        if (isinstance<cell_connection_shim>(c)) {
+            cons.push_back(cast<cell_connection_shim>(c));
+        }
+        else {
+            throw std::runtime_error(
```
Comment (akuesters) [inline python/recipe.cpp]: Why not use `pyarb_error` as in `event_generators`?

File: python/recipe.cpp [hunk 2]
```diff
@@ -133,7 +165,7 @@ void register_recipe(pybind11::module& m) {
         "  Defined by source and destination end points (that is pre-synaptic and post-synaptic respectively), a connection weight and a delay time.");
     cell_connection
         .def(pybind11::init<arb::cell_member_type, arb::cell_member_type, float, arb::time_type>(),
-            "source"_a = arb::cell_member_type{0,0}, "dest"_a = arb::cell_member_type{0,0}, "weight"_a = 0.f, "delay"_a,
+            "source"_a, "dest"_a, "weight"_a, "delay"_a,
             "Construct a connection with arguments:\n"
             "  source:      The source end point of the connection (default (0,0)).\n"
```
Comment (akuesters) [inline python/recipe.cpp]: We should also delete the defaults in descriptions and make unit description consistent (e.g. in lif_cell only [] are used, here unit: ).

File: python/simulation.cpp [hunk 3]
```diff
@@ -33,9 +33,9 @@ void register_simulation(pybind11::module& m) {
         .def("run", &arb::simulation::run,
             pybind11::call_guard<pybind11::gil_scoped_release>(),
             "Run the simulation from current simulation time to tfinal, with maximum time step size dt.",
-            "tfinal"_a, "dt"_a)
+            "tfinal"_a, "dt"_a=0.025)
```
Comment (akuesters) [inline python/simulation.cpp]: Maybe add default and units (of dt and tfinal) to description?

File: python/recipe.hpp [hunk 4]
```diff
@@ -39,17 +39,16 @@ class py_recipe {
...
-    virtual std::vector<pybind11::object> event_generators(arb::cell_gid_type gid) const {
-        auto guard = pybind11::gil_scoped_acquire();
-        return {};
-    }
+    virtual std::vector<pybind11::object> event_generators(arb::cell_gid_type gid) const { return {}; }
 
-    virtual std::vector<arb::cell_connection> connections_on(arb::cell_gid_type gid) const { return {}; }
+    //virtual std::vector<arb::cell_connection> connections_on(arb::cell_gid_type gid) const { return {}; }
+    virtual std::vector<pybind11::object> connections_on(arb::cell_gid_type gid) const { return {}; }
     virtual std::vector<arb::gap_junction_connection> gap_junctions_on(arb::cell_gid_type) const { return {}; }
```
[the remark this answers is not in the document]
Comment (bcumming) [reply python/recipe.hpp]: I have removed the connection shim, to simplify conversion from Python to C++ code.

File: python/recipe.cpp (hunk 1 above)
Comment (bcumming) [reply python/recipe.cpp]: Fixed.

File: python/simulation.cpp (hunk 3 above)
Comment (bcumming) [reply python/simulation.cpp]: I have added the unit here, however these further changes are beyond the scope of this PR.

**Pull request merged**